### DISCUSSION TITLE
Is `flux diff` supposed to work with postBuild?

### DISCUSSION THREAD

**Discussion by dmitriishaburov** [fluxcd/flux2#4117] [Q&A]
I want to add env substitution, but before that I planned to check if I have some unexpected diffs from enabling it.
But it seems that `flux diff` is completely ignoring it, is that intended behaviour?

```
► Kustomization/flux-system/flux-system drifted
spec
  + one map entry added:
    postBuild:
      substituteFrom:
      - name: eks
        kind: ConfigMap
        optional: false

► HelmRelease/kube-system/aws-ebs-csi-driver drifted
spec.values.controller.serviceAccount.annotations.eks.amazonaws.com/role-arn
  ± value change
    - arn:aws:iam::111222333444:role/eks-my-cluster-ebs-csi
    + arn:aws:iam::${EKS_ACCOUNT_ID}:role/eks-${EKS_CLUSTER_NAME}-ebs-csi
```

Expected behaviour: see substituted variables in `aws-ebs-csi-driver`

**Comment by stefanprodan** [comment]
> But it seems that flux diff is completely ignoring it, is that intended behaviour?

Yes, only inline vars are taking into account.

**Reply by dmitriishaburov** [reply]
Hm, did I understood `inline vars` wrong? Tried with just `substitute`, no difference:
```
► Kustomization/flux-system/flux-system drifted
spec
  + one map entry added:
    postBuild:
      substitute:
        EKS_ACCOUNT_ID: 111222333444
        EKS_CLUSTER_NAME: my-cluster

► HelmRelease/kube-system/aws-ebs-csi-driver drifted
spec.values.controller.serviceAccount.annotations.eks.amazonaws.com/role-arn
  ± value change
    - arn:aws:iam::111222333444:role/eks-my-cluster-ebs-csi
    + arn:aws:iam::${EKS_ACCOUNT_ID}:role/eks-${EKS_CLUSTER_NAME}-ebs-csi
```

**Reply by stefanprodan** [reply]
Did you used `--kustomization-file` and are you on the latest version?

**Reply by dmitriishaburov** [reply]
```
$ flux version
flux: v2.0.1
helm-controller: v0.35.0
kustomize-controller: v1.0.1
source-controller: v1.0.1

$ flux diff kustomization flux-system --path . --kustomization-file ./flux-system/gotk-sync.yaml
```

**Reply by stefanprodan** [reply]
Hmm why are you modifying `gotk-sync.yaml` that file will get overwritten by flux bootstrat. All the cluster addons should be managed by a dedicated Flux Kustomization, please see here how to properly structure your repo: https://github.com/fluxcd/flux2-kustomize-helm-example

**Reply by dmitriishaburov** [reply]
Ok, managed to make substitution work locally with `flux diff`, but in a very strange way:
- I'm using default `gotk-sync.yaml` which have GitRepository and Kustomization
- I'm adding `postBuild` section via patch in `flux-system/kustomization.yaml`

In order to `flux diff` work I've had to:
- Remove GitRepository from `gotk-sync.yaml` so it'd have only Kustomization
- Move postBuild section from patch directly to `gotk-sync.yaml`

That seems like a bug tbh or really strange behaviour

**Reply by stefanprodan** [reply]
You’re not supposed to alter the flux-system kustomization but instead create a dedicated one for clusters addons, apps, etc.

**Reply by dmitriishaburov** [reply]
I see the point. Problem would most likely still stay though with separate kustomization, if it not first object in yaml file or have `postBuild` added via kustomize.
